### PR TITLE
[Date Range Input] - Part 7: Typing dates into input fields

### PR DESCRIPTION
**Pull request by cmslewis**
#### Checklist

- [x] Include tests

#### Changes proposed in this pull request:

- _New feature:_ Can now type dates into the input fields, and the picker selection will update (if uncontrolled).
- _Bugfix:_ Fix parallel issue I noticed in `DateInput` where editing the input field in controlled mode overwrote a controlled value. Now, if the user changes the text in an input field in controlled mode, (1) the calendar selection will not change, and (2) the controlled value will be repopulated in the field on blur. (And to be clear, I fixed this just for `DateRangeInput`, not for `DateInput` too).

_Changes NOT included:_
- **I didn't implement input validation yet**. It's perfectly possible to input backward or otherwise invalid date ranges. The next PR will deal with input validation.
- **I didn't implement focus-handling yet.** The focused field has no bearing on which end of the date range will be specified by the next click. A near-future PR will deal with focus handling. 

#### Reviewers should focus on:

- What's the best way to support generic `onBlur`, `onFocus`, and `onChange` handlers for the two input fields, given that each field needs to process the same logic with different pieces of `state`?

    In my prototype from a couple weeks ago, I passed boundary-specific `state` keys (e.g. `"isStartInputFocused"`, `"selectedEnd"`) as separate string parameters from `handleStartInputBlur` (e.g.) into `handleGenericInputBlur` (e.g.).

    The `IStateKeysAndValuesObject` here looks a little smelly at a glance, but it feels nicer than my previous approach.

    Or maybe it's better to pull the logic out into a couple child components (called `DateRangeInputField`?) Not sure how much mileage that'll get us, but happy to take a look if we're not okay with the current implementation.

- How does the controlled-mode _Bugfix_ sound?

#### Screenshot

_Uncontrolled Mode_

_Controlled Mode_

### CODE REVIEW

**Comment by cmslewis** [comment]
Update: I'm playing with pulling input-field logic into a separate component. Will report back with results.

**Comment by llorca** [comment]
This may not be for this PR, but note that the input loses focus after selecting a start date

**Comment by cmslewis** [comment]
@llorca - yep, will handle focus-state management in a future PR. 👍

File: packages/datetime/src/dateRangeInput.tsx [hunk 1]
```diff
@@ -190,8 +221,114 @@ export class DateRangeInput extends AbstractComponent<IDateRangeInputProps, IDat
...
+
+    private handleEndInputBlur = (e: React.FormEvent<HTMLInputElement>) => {
+        this.handleInputBlur(e, DateRangeBoundary.END);
+    }
+
+    private handleInputBlur = (_e: React.FormEvent<HTMLInputElement>, boundary: DateRangeBoundary) => {
+        const { keys, values } = this.getStateKeysAndValuesForBoundary(boundary);
+
+        const maybeNextValue = this.dateStringToMoment(values.inputString);
+        const isValueControlled = this.props.value !== undefined;
```
Comment (leebyp) [inline packages/datetime/src/dateRangeInput.tsx]: fan of starting to use a private method to keep track of this across our components - started doing this here: https://github.com/palantir/blueprint/blob/master/packages/datetime/src/dateRangePicker.tsx#L101

thoughts?

File: packages/datetime/src/dateRangeInput.tsx [hunk 2]
```diff
@@ -190,8 +221,114 @@ export class DateRangeInput extends AbstractComponent<IDateRangeInputProps, IDat
...
+
+        if (values.inputString == null || values.inputString.length === 0) {
+            if (isValueControlled) {
+                this.setState({
+                    [keys.isInputFocused]: false,
+                    [keys.inputString]: this.getFormattedDateString(values.controlledValue),
```
Comment (leebyp) [inline packages/datetime/src/dateRangeInput.tsx]: yup, i agree with this interaction

File: packages/datetime/src/dateRangeInput.tsx [hunk 3]
```diff
@@ -211,17 +348,74 @@ export class DateRangeInput extends AbstractComponent<IDateRangeInputProps, IDat
...
+                    inputString: "endInputString",
+                    isInputFocused: "isEndInputFocused",
+                    selectedValue: "selectedEnd",
+                },
+                values: {
+                    controlledValue: (controlledValue != null) ? controlledValue[1] : undefined,
+                    inputString: this.state.endInputString,
+                    isFocused: this.state.isEndInputFocused,
+                    selectedValue: this.state.selectedEnd,
+                },
```
Comment (leebyp) [inline packages/datetime/src/dateRangeInput.tsx]: The `DateRangeInputField` you mention in the description will be another step towards keeping consistencies between `DateInput` and `DateRangeInput` though I am wary we might get into another rabbit hole

File: packages/datetime/src/dateRangeInput.tsx [hunk 4]
```diff
@@ -81,10 +82,36 @@ export interface IDateRangeInputProps extends IDatePickerBaseProps, IProps {
...
+    startInputString?: string;
+    endInputString?: string;
+
     selectedEnd?: moment.Moment;
     selectedStart?: moment.Moment;
 };
 
+interface IStateKeysAndValuesObject {
+    keys: {
+        inputString: string;
```
Comment (leebyp) [inline packages/datetime/src/dateRangeInput.tsx]: i think you can put the actual strings here to enforce the type

File: packages/datetime/src/dateRangeInput.tsx (hunk 3 above)
Comment (leebyp) [inline packages/datetime/src/dateRangeInput.tsx]: even if we split it out, will this parent still need to utilise a lot of the internals from the hypothetical new component? and the code be just as complicated?

File: packages/datetime/src/dateRangeInput.tsx (hunk 1 above)
Comment (cmslewis) [inline packages/datetime/src/dateRangeInput.tsx]: Yeah, I thought about that too. We should just be careful to note that `null` may be a meaningful value for `this.props.value`, so not every `isControlled()` implementation will be the same (some will need a strict `!== undefined` check).

File: packages/datetime/src/dateRangeInput.tsx (hunk 3 above)
Comment (cmslewis) [inline packages/datetime/src/dateRangeInput.tsx]: Update: I tried splitting each field out into a common, fully-controlled component. The LOC decreased in `dateRangeInput.tsx`, but things actually got hairier and less clear, plus there's still some boilerplate with respect to updating the appropriate state values in the parent component. Wasn't loving it.

File: packages/datetime/src/dateRangeInput.tsx (hunk 4 above)
Comment (cmslewis) [inline packages/datetime/src/dateRangeInput.tsx]: What do you mean exactly?

**Pull request merged**